src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -26,10 +27,21 @@ public IEnumerator<TElement> GetEnumerator()
             Buffer<TElement> buffer = new Buffer<TElement>(_source);
             if (buffer._count > 0)
             {
-                int[] map = SortedMap(buffer);
-                for (int i = 0; i < buffer._count; i++)
+                int[] map = null;
+                try
                 {
-                    yield return buffer._items[map[i]];
+                    map = SortedMap(buffer);
+                    for (int i = 0; i < buffer._count; i++)
+                    {
+                        yield return buffer._items[map[i]];
+                    }
+                }
+                finally
```

Comment:
The corruption issue might be relevant even without finally. That is really an issue of being possible to resume an iterator on multiple threads. 
It would be very bad thing to do, but would we make it worse by exposing shared pool to that as well.